content/blog/2021-11-10-agent.md

```diff
@@ -0,0 +1,156 @@
+
+![Prometheus agent](/assets/blog/2021-11-10/agent.png)
+
+The best part about Prometheus Agent is that it's built into Prometheus. Same scraping APIs, same semantics, same configuration and discovery mechanism.
+
+What benefits of using Agent mode if you plan not to query or alert on data locally and stream metrics outside? There are a few:
+
+First of all, efficiency. Our customized Agent TSDB WAL removes the data immediately after successful writes. If it cannot reach the remote endpoint, it persists the data temporarily on the disk until the remote endpoint is back online. This is currently limited to 2h buffer only, similar to non-agent Prometheus, [hopefully unblocked soon](https://github.com/prometheus/prometheus/issues/9607). This means that we don't need to build chunks of data in memory. We don't need to maintain a full index for querying purposes. Essentially Agent uses a fraction of resources normal Prometheus would use in a similar situation.
+
+Does this efficiency matter? Yes! As we mentioned, every GB of memory and every CPU core used on the edge clusters matters for some deployments. On the other hand, the paradigm of performing monitoring using metrics is quite mature these days. This means that the more relevant metrics with more cardinality you can ship for the same cost - the better.
+
+> NOTE: With the introduction of the Agent Mode, the original Prometheus server mode still stays as the recommended, stable and maintained mode. Agent Mode with remote storage brings additional complexity. Use with care.
+
+Secondly, the benefit of the new Agent mode is that it enables easier ingestion horizontal scalability. This is something I am excited about the most. Let me explain why.
+
+### The Dream: Auto-Scalable Metric Ingestion
+
+The true auto-scaling solution for scraping would be based on the amount of metric target and the number of metrics they expose. The more data we have to scrape, the more instances of Prometheus we deploy automatically. If the number of targets or its data goes down, we could scale down and remove a couple of instances. This would remove the manual burden of adjusting the sizing of Prometheus and stop the need to over-allocating Prometheus for situations where the cluster is temporarily small.
+
+With just Prometheus in server mode, this was hard to achieve. This is because Prometheus in server mode is stateful. Whatever is collected stays as-is in a single place. This means that the scale-down procedure would need to back up the collected data to existing instances before termination. Then we would have the problem of overlapped scraped, misleading staleness markers etc.
```

Comment:
```suggestion
With just Prometheus in server mode, this was hard to achieve. This is because Prometheus in server mode is stateful. Whatever is collected stays as-is in a single place. This means that the scale-down procedure would need to back up the collected data to existing instances before termination. Then we would have the problem of overlapping scrapes, misleading staleness markers etc.
```